plugin/caffe/caffe_data_iter.cc

```diff
@@ -0,0 +1,217 @@
+/*!
+ * Copyright (c) 2015 by Contributors
+ * \file caffe_data_iter.cc
+ * \brief register mnist iterator
+*/
+#include <caffe/proto/caffe.pb.h>
+#include <dmlc/logging.h>
+#include <dmlc/parameter.h>
+#include <mxnet/operator.h>
+#include <atomic>
+
+#include "../../src/operator/operator_common.h"
+#include "caffe_common.h"
+#include "caffe_stream.h"
+#include "caffe_fieldentry.h"
+#include "caffe_blob.h"
+#include "../../src/io/inst_vector.h"
+#include "../../src/io/iter_prefetcher.h"
+#include <../../src/operator/cast-inl.h>
+
+namespace mxnet {
+namespace io {
+
+struct CaffeDataParam : public dmlc::Parameter<CaffeDataParam> {
+  /*! \brief protobuf text */
+  ::caffe::LayerParameter prototxt;
+  /*! \brief maximum iteration count */
+  int max_iterations;  
+  /*! \brief data mode */
+  bool flat;
+
+  DMLC_DECLARE_PARAMETER(CaffeDataParam) {
+    DMLC_DECLARE_FIELD(prototxt).set_default("layer{}")
+      .describe("Caffe's layer parameter");
+    DMLC_DECLARE_FIELD(flat).set_default(false)
+      .describe("Augmentation Param: Whether to flat the data into 1D.");
+    DMLC_DECLARE_FIELD(max_iterations).set_lower_bound(1).set_default(10000)
+      .describe("Maximum number of iterations per epoch.");
+  }
+};
+
+struct CaffePrefetchParam : dmlc::Parameter<CaffePrefetchParam> {
+    /*! \brief data type */
+    int dtype;
+    
+    DMLC_DECLARE_PARAMETER(CaffePrefetchParam) {
+      DMLC_DECLARE_FIELD(dtype)
+        .add_enum("float32", mshadow::kFloat32)
+        .add_enum("float64", mshadow::kFloat64)
+        .add_enum("float16", mshadow::kFloat16)
+        .add_enum("uint8", mshadow::kUint8)
+        .add_enum("int32", mshadow::kInt32)
+        .set_default(mshadow::default_type_flag)
+        .describe("Data type.");
+    }
+};
+
+template<typename Dtype>
+class CaffeDataIter : public IIterator<TBlobBatch> {
+
+public:
+  CaffeDataIter(void) : loc_(0), batch_size_(0), channels_(0), width_(0), height_(0) {}
+  virtual ~CaffeDataIter(void) {}
+    
+  // intialize iterator loads data in
+  virtual void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) {
+    std::map<std::string, std::string> kmap(kwargs.begin(), kwargs.end());
+    param_.InitAllowUnknown(kmap);
```

Comment:
Lenet_train_test.prototxt has the phase inside an "include" block rather
than at root level of the layer. Caffe seems to understand this.
On Mon, Aug 22, 2016 at 5:18 PM Chris Olivier <cjolivier01@gmail.com> wrote:

> Blindly copier allow unknown from similar class. My assumption was that it
> doesn't error if a parameter isn't part of that particular parameter class,
> which would be desired in this class since those parameters are parsed by
> three separate parameter objects (prefrtchiter this class and the type
> parameter class )
> On Mon, Aug 22, 2016 at 5:14 PM Eric Junyuan Xie <notifications@github.com>
> wrote:
>
>> In plugin/caffe/caffe_data_iter.cc
>> <https://github.com/dmlc/mxnet/pull/3103#discussion_r75783277>:
>>
>> > +        .set_default(mshadow::default_type_flag)
>> > +        .describe("Data type.");
>> > +    }
>> > +};
>> > +
>> > +template<typename Dtype>
>> > +class CaffeDataIter : public IIterator<TBlobBatch> {
>> > +
>> > +public:
>> > +  CaffeDataIter(void) : loc_(0), batch_size_(0), channels_(0), width_(0), height_(0) {}
>> > +  virtual ~CaffeDataIter(void) {}
>> > +
>> > +  // intialize iterator loads data in
>> > +  virtual void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) {
>> > +    std::map<std::string, std::string> kmap(kwargs.begin(), kwargs.end());
>> > +    param_.InitAllowUnknown(kmap);
>>
>> Any reason for using allowunknown?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/dmlc/mxnet/pull/3103/files/e780973a75910edacd28aec003af6a3f6bf8213e#r75783277>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AKts_euOerJFD62whCc6tba4o7cLknWyks5qijt1gaJpZM4JqHuN>
>> .
>>
>